File: LeetCode/232用栈实现队列.go
type MyQueue struct {
	in, out []int
}

func Constructor() MyQueue {
	return MyQueue{}
}

func (this *MyQueue) Push(x int) {
	this.in = append(this.in, x)
}

func (this *MyQueue) inToOut() {
	for len(this.in) > 0 {
		this.out = append(this.out, this.in[len(this.in)-1])
		this.in = this.in[:len(this.in)-1]
	}
}

func (this *MyQueue) Pop() int {
	if len(this.out) == 0 {
		this.inToOut()
	}
	x := this.out[len(this.out)-1]
	this.out = this.out[:len(this.out)-1]
	return x
}

func (this *MyQueue) Peek() int {
	if len(this.out) == 0 {
		this.inToOut()
	}
	return this.out[len(this.out)-1]
}

func (this *MyQueue) Empty() bool {
	return len(this.in) == 0 && len(this.out) == 0
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */

File: LeetCode/232用栈实现队列.java
class MyQueue {
    Stack<Integer> staIn;
    Stack<Integer> staOut;

    public MyQueue() {
        staIn = new Stack<>();
        staOut = new Stack<>();
    }
    
    public void push(int x) {
        staIn.push(x);
    }
    
    public int pop() {
        copyIntoOut();
        return staOut.pop();
    }
    
    public int peek() {
        copyIntoOut();
        return staOut.peek();
    }
    
    public boolean empty() {
        return staIn.isEmpty() && staOut.isEmpty();
    }

    public void copyIntoOut() {
        if (!staOut.isEmpty()) return ;
        while (!staIn.isEmpty()) {
            staOut.push(staIn.pop());
        }
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */